// backport/dist/services/github/v4/fetchPullRequestBySearchQuery.private.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const getDevAccessToken_1 = require("../../../test/private/getDevAccessToken");
const fetchPullRequestBySearchQuery_1 = require("./fetchPullRequestBySearchQuery");
describe('fetchPullRequestBySearchQuery', () => {
    let devAccessToken;
    beforeAll(async () => {
        devAccessToken = await getDevAccessToken_1.getDevAccessToken();
    });
    describe('when filter does not match any PRs', () => {
        it('throws an error', async () => {
            const options = {
                accessToken: devAccessToken,
                all: false,
                author: 'sqren',
                githubApiBaseUrlV4: 'https://api.github.com/graphql',
                maxNumber: 10,
                prFilter: 'label:non-existing',
                repoName: 'backport-e2e',
                repoOwner: 'backport-org',
                sourceBranch: 'master',
            };
            await expect(fetchPullRequestBySearchQuery_1.fetchPullRequestBySearchQuery(options)).rejects.toThrowError('There are no commits by "sqren" matching the filter "label:non-existing". Try with `--all` for commits by all users or `--author=<username>` for commits from a specific user');
        });
    });
    describe('when filter matches PRs', () => {
        it('returns the merge commits for those PRs', async () => {
            const options = {
                accessToken: devAccessToken,
                all: false,
                author: 'sqren',
                githubApiBaseUrlV4: 'https://api.github.com/graphql',
                maxNumber: 10,
                prFilter: 'label:v7.8.0',
                repoName: 'backport-e2e',
                repoOwner: 'backport-org',
                sourceBranch: 'master',
            };
            expect(await fetchPullRequestBySearchQuery_1.fetchPullRequestBySearchQuery(options)).toEqual([
                {
                    existingTargetPullRequests: [],
                    formattedMessage: 'Add sheep emoji (#9)',
                    pullNumber: 9,
                    sha: 'eebf165c82a4b718d95c11b3877e365b1949ff28',
                    sourceBranch: 'master',
                    targetBranchesFromLabels: [],
                },
                {
                    existingTargetPullRequests: [],
                    formattedMessage: 'Add 🍏 emoji (#5)',
                    pullNumber: 5,
                    sha: 'ee8c492334cef1ca077a56addb79a26f79821d2f',
                    sourceBranch: 'master',
                    targetBranchesFromLabels: [],
                },
            ]);
        });
    });
});
