// backport/dist/services/github/v4/fetchExistingPullRequest.private.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const getDevAccessToken_1 = require("../../../test/private/getDevAccessToken");
const fetchExistingPullRequest_1 = require("./fetchExistingPullRequest");
describe('fetchExistingPullRequest', () => {
    let devAccessToken;
    beforeAll(async () => {
        devAccessToken = await getDevAccessToken_1.getDevAccessToken();
    });
    describe('when PR does not exist', () => {
        it('returns undefined', async () => {
            const options = {
                accessToken: devAccessToken,
                githubApiBaseUrlV4: 'https://api.github.com/graphql',
            };
            const prPayload = {
                owner: 'backport-org',
                repo: 'backport-e2e',
                title: 'My PR title',
                body: 'My PR body',
                head: 'sqren:backport/7.8/pr-foo',
                base: '7.8',
            };
            const res = await fetchExistingPullRequest_1.fetchExistingPullRequest({ options, prPayload });
            expect(res).toBe(undefined);
        });
    });
    describe('when PR exists', () => {
        it('returns the PR number and url', async () => {
            const options = {
                accessToken: devAccessToken,
                githubApiBaseUrlV4: 'https://api.github.com/graphql',
            };
            const prPayload = {
                owner: 'backport-org',
                repo: 'backport-e2e',
                title: 'My PR title',
                body: 'My PR body',
                head: 'sqren:backport/7.8/pr-9',
                base: '7.8',
            };
            const res = await fetchExistingPullRequest_1.fetchExistingPullRequest({ options, prPayload });
            expect(res).toEqual({
                url: 'https://github.com/backport-org/backport-e2e/pull/10',
                number: 10,
            });
        });
    });
});
