// backport/dist/services/github/v4/mocks/getPullRequestEdgeMock.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getPullRequestEdgeMock = void 0;
function getPullRequestEdgeMock({ pullRequestNumber, timelinePullRequest, }) {
    return {
        node: {
            labels: {
                nodes: [{ name: 'my-label-a' }],
            },
            mergeCommit: {
                oid: 'f3b618b9421fdecdb36862f907afbdd6344b361d',
            },
            repository: {
                name: 'kibana',
                owner: {
                    login: 'elastic',
                },
            },
            number: pullRequestNumber,
            timelineItems: {
                edges: [
                    {
                        node: {
                            source: {
                                __typename: 'PullRequest',
                                title: timelinePullRequest.title,
                                state: 'MERGED',
                                commits: {
                                    edges: timelinePullRequest.commits.map((message) => ({
                                        node: { commit: { message } },
                                    })),
                                },
                                baseRefName: '7.x',
                            },
                        },
                    },
                ],
            },
        },
    };
}
exports.getPullRequestEdgeMock = getPullRequestEdgeMock;
