// backport/dist/services/github/v3/addAssigneesToPullRequest.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.addAssigneesToPullRequest = void 0;
const rest_1 = require("@octokit/rest");
const ora_1 = __importDefault(require("ora"));
const logger_1 = require("../../logger");
async function addAssigneesToPullRequest({ githubApiBaseUrlV3, repoName, repoOwner, accessToken, username, dryRun, }, pullNumber, assignees) {
    const isSelfAssigning = assignees.length === 1 && assignees[0] === username;
    const text = isSelfAssigning
        ? `Self-assigning to #${pullNumber}`
        : `Adding assignees to #${pullNumber}: ${assignees.join(', ')}`;
    logger_1.logger.info(text);
    const spinner = ora_1.default(text).start();
    try {
        if (dryRun) {
            spinner.succeed(`Dry run: ${text}`);
            return;
        }
        const octokit = new rest_1.Octokit({
            auth: accessToken,
            baseUrl: githubApiBaseUrlV3,
            log: logger_1.logger,
        });
        await octokit.issues.addAssignees({
            owner: repoOwner,
            repo: repoName,
            issue_number: pullNumber,
            assignees: assignees,
        });
        spinner.succeed();
    }
    catch (e) {
        spinner.fail();
        logger_1.logger.info(`Could not add assignees to PR ${pullNumber}`, e.stack);
    }
}
exports.addAssigneesToPullRequest = addAssigneesToPullRequest;
