// backport/dist/services/github/v3/addAssigneesToPullRequest.test.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const nock_1 = __importDefault(require("nock"));
const addAssigneesToPullRequest_1 = require("./addAssigneesToPullRequest");
describe('addAssigneesToPullRequest', () => {
    it('should add assignees to PR', async () => {
        const pullNumber = 216;
        const assignees = ['sqren'];
        const scope = nock_1.default('https://api.github.com')
            .post('/repos/backport-org/backport-demo/issues/216/assignees', {
            assignees: ['sqren'],
        })
            .reply(200, 'some response');
        const res = await addAssigneesToPullRequest_1.addAssigneesToPullRequest({
            githubApiBaseUrlV3: 'https://api.github.com',
            repoName: 'backport-demo',
            repoOwner: 'backport-org',
            accessToken: 'my-token',
            username: 'sqren',
            dryRun: false,
        }, pullNumber, assignees);
        expect(res).toBe(undefined);
        scope.done();
        nock_1.default.cleanAll();
    });
});
