// backport/dist/services/github/v3/addLabelsToPullRequest.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.addLabelsToPullRequest = void 0;
const rest_1 = require("@octokit/rest");
const ora_1 = __importDefault(require("ora"));
const logger_1 = require("../../logger");
async function addLabelsToPullRequest({ githubApiBaseUrlV3, repoName, repoOwner, accessToken, dryRun, }, pullNumber, labels) {
    const text = `Adding labels: ${labels.join(', ')}`;
    logger_1.logger.info(text);
    const spinner = ora_1.default(text).start();
    try {
        if (dryRun) {
            spinner.succeed(`Dry run: ${text}`);
            return;
        }
        const octokit = new rest_1.Octokit({
            auth: accessToken,
            baseUrl: githubApiBaseUrlV3,
            log: logger_1.logger,
        });
        await octokit.issues.addLabels({
            owner: repoOwner,
            repo: repoName,
            issue_number: pullNumber,
            labels,
        });
        spinner.succeed();
    }
    catch (e) {
        spinner.fail();
        logger_1.logger.info(`Could not add labels to PR ${pullNumber}`, e.stack);
    }
}
exports.addLabelsToPullRequest = addLabelsToPullRequest;
