// shared-github-internals/lib/tests/context.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const Octokit = require("@octokit/rest");
const envalid = require("envalid");
const createTestContext = () => {
    const env = envalid.cleanEnv(
    // eslint-disable-next-line no-process-env
    process.env, {
        GITHUB_PERSONAL_ACCESS_TOKEN: envalid.str({
            desc: "The token must grant read/write access to the test repository.",
            docs: "https://github.com/settings/tokens",
        }),
        GITHUB_TEST_REPOSITORY_NAME: envalid.str({
            desc: "Name of the repository against which the tests will be run",
        }),
        GITHUB_TEST_REPOSITORY_OWNER: envalid.str({
            desc: "Owner of the repository against which the tests will be run.",
        }),
    }, { strict: true });
    return {
        octokit: new Octokit({
            auth: `token ${env.GITHUB_PERSONAL_ACCESS_TOKEN}`,
        }),
        owner: env.GITHUB_TEST_REPOSITORY_OWNER,
        repo: env.GITHUB_TEST_REPOSITORY_NAME,
    };
};
exports.createTestContext = createTestContext;
